File: python/nexuskv/contracts/generated.py
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

SCHEMA_VERSION = "nexuskv.contract.v1"

class EngineFamily(StrEnum):
    UNKNOWN = "unknown"
    SGLANG = "sglang"
    VLLM = "vllm"

class StateSemanticType(StrEnum):
    MHA_KV = "mha_kv"
    GQA_KV = "gqa_kv"
    MQA_KV = "mqa_kv"
    MLA_STATE = "mla_state"
    GENERIC_CONTAINER = "generic_container"

class TensorRole(StrEnum):
    KEY = "key"
    VALUE = "value"
    LATENT = "latent"
    POSITION = "position"
    AUXILIARY = "auxiliary"

class Granularity(StrEnum):
    TOKEN = "token"
    BLOCK = "block"
    PAGE = "page"
    SEGMENT = "segment"

class CompatibilityFlag(StrEnum):
    EXACT_REUSE = "exact_reuse"
    PREFIX_REUSE = "prefix_reuse"
    PAGE_REUSE = "page_reuse"
    BLOCK_REUSE = "block_reuse"
    WARM_START = "warm_start"
    POST_EVICTION_RELOAD = "post_eviction_reload"

class DeviceClass(StrEnum):
    CPU = "cpu"
    CUDA = "cuda"
    ROCM = "rocm"
    TPU = "tpu"
    OTHER = "other"

class BufferKind(StrEnum):
    DEVICE = "device"
    HOST_PINNED = "host_pinned"
    HOST_PAGEABLE = "host_pageable"
    REMOTE = "remote"
    FILE_BACKED = "file_backed"

class TransferBackend(StrEnum):
    BASELINE_TRANSPORT = "baseline_transport"
    STAGED_COPY = "staged_copy"
    RDMA = "rdma"
    ZERO_COPY = "zero_copy"

class TransferCapability(StrEnum):
    HOST_TO_DEVICE = "host_to_device"
    DEVICE_TO_HOST = "device_to_host"
    HOST_TO_STORE = "host_to_store"
    STORE_TO_HOST = "store_to_host"
    ASYNC = "async"
    ZERO_COPY_CANDIDATE = "zero_copy_candidate"

class TierKind(StrEnum):
    DEVICE = "device"
    HOST_DRAM = "host_dram"
    LOCAL_SSD = "local_ssd"
    REMOTE_SHARED = "remote_shared"
    OBJECT_STORE = "object_store"

class MaterializationCapability(StrEnum):
    FULL = "full"
    PARTIAL = "partial"
    PREFETCH = "prefetch"
    ASYNC_FETCH = "async_fetch"
    FALLBACK_RECOMPUTE = "fallback_recompute"

class MatchClassification(StrEnum):
    EXACT = "exact"
    PREFIX = "prefix"
    PARTIAL = "partial"

class PlanDisposition(StrEnum):
    FULL_REUSE = "full_reuse"
    PARTIAL_REUSE = "partial_reuse"
    RECOMPUTE = "recompute"

@dataclass(slots=True)
class TensorSpec:
    name: str
    role: TensorRole
    dtype: str
    shape: list[str]

@dataclass(slots=True)
class QuantizationMetadata:
    scheme: str
    bits: int
    group_size: int

@dataclass(slots=True)
class LayoutMetadata:
    layout: str
    page_tokens: int
    block_tokens: int
    packed: bool

@dataclass(slots=True)
class TransferPath:
    backend: TransferBackend
    capabilities: list[TransferCapability]

@dataclass(slots=True)
class MaterializationProfile:
    capabilities: list[MaterializationCapability]
    tier_kinds: list[TierKind]
    device_classes: list[DeviceClass]
    buffer_kinds: list[BufferKind]

@dataclass(slots=True)
class AttentionStateDescriptor:
    schema_version: str
    descriptor_id: str
    engine_family: EngineFamily
    semantic_type: StateSemanticType
    granularity: Granularity
    tensor_specs: list[TensorSpec]
    quantization: QuantizationMetadata
    layout: LayoutMetadata
    compatibility_flags: list[CompatibilityFlag]
    transfer_paths: list[TransferPath]
    materialization: MaterializationProfile
    layout_metadata: dict[str, str]

@dataclass(slots=True)
class KeyIdentity:
    tenant: str
    namespace: str
    model: str
    engine_family: EngineFamily
    semantic_type: StateSemanticType
    tokens: list[int]
    block_id: int | None
    page_id: int | None

@dataclass(slots=True)
class ReuseKey:
    identity: KeyIdentity

@dataclass(slots=True)
class QueryKey:
    identity: KeyIdentity

@dataclass(slots=True)
class EntryVersion:
    generation: int
    lineage: str

@dataclass(slots=True)
class EntryLocation:
    tier: TierKind
    locator: str

@dataclass(slots=True)
class PolicyHint:
    reusable: bool
    admission_hint: str
    eviction_hint: str

@dataclass(slots=True)
class EntryIdentity:
    key: KeyIdentity
    entry_id: str
    version: EntryVersion

@dataclass(slots=True)
class CacheEntry:
    identity: EntryIdentity
    descriptor: AttentionStateDescriptor
    location: EntryLocation
    policy_hint: PolicyHint

@dataclass(slots=True)
class MatchExtent:
    units: int
    granularity: Granularity

@dataclass(slots=True)
class RemainingWork:
    tokens: list[int]
    fetch_required: bool
    recompute_required: bool

@dataclass(slots=True)
class CompatibilitySignal:
    reusable: bool
    fallback_to_recompute: bool
    reason: str

@dataclass(slots=True)
class MatchResult:
    classification: MatchClassification
    matched_key: ReuseKey
    requested_key: QueryKey
    matched_extent: MatchExtent
    entry: CacheEntry
    remaining: RemainingWork
    compatibility: CompatibilitySignal

@dataclass(slots=True)
class ReusableSlice:
    tokens: list[int]
    source_tier: TierKind

@dataclass(slots=True)
class PartialHitPlan:
    disposition: PlanDisposition
    reusable: ReusableSlice
    remaining: RemainingWork
    entry: CacheEntry

__all__ = [
    "SCHEMA_VERSION",
    "EngineFamily",
    "StateSemanticType",
    "TensorRole",
    "Granularity",
    "CompatibilityFlag",
    "DeviceClass",
    "BufferKind",
    "TransferBackend",
    "TransferCapability",
    "TierKind",
    "MaterializationCapability",
    "MatchClassification",
    "PlanDisposition",
    "TensorSpec",
    "QuantizationMetadata",
    "LayoutMetadata",
    "TransferPath",
    "MaterializationProfile",
    "AttentionStateDescriptor",
    "KeyIdentity",
    "ReuseKey",
    "QueryKey",
    "EntryVersion",
    "EntryLocation",
    "PolicyHint",
    "EntryIdentity",
    "CacheEntry",
    "MatchExtent",
    "RemainingWork",
    "CompatibilitySignal",
    "MatchResult",
    "ReusableSlice",
    "PartialHitPlan",
]

File: rust/crates/nexus-state/src/generated.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

pub const SCHEMA_VERSION: &str = "nexuskv.contract.v1";

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EngineFamily {
    Unknown,
    Sglang,
    Vllm,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StateSemanticType {
    MhaKv,
    GqaKv,
    MqaKv,
    MlaState,
    GenericContainer,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TensorRole {
    Key,
    Value,
    Latent,
    Position,
    Auxiliary,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Granularity {
    Token,
    Block,
    Page,
    Segment,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CompatibilityFlag {
    ExactReuse,
    PrefixReuse,
    PageReuse,
    BlockReuse,
    WarmStart,
    PostEvictionReload,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DeviceClass {
    Cpu,
    Cuda,
    Rocm,
    Tpu,
    Other,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BufferKind {
    Device,
    HostPinned,
    HostPageable,
    Remote,
    FileBacked,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TransferBackend {
    BaselineTransport,
    StagedCopy,
    Rdma,
    ZeroCopy,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TransferCapability {
    HostToDevice,
    DeviceToHost,
    HostToStore,
    StoreToHost,
    Async,
    ZeroCopyCandidate,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TierKind {
    Device,
    HostDram,
    LocalSsd,
    RemoteShared,
    ObjectStore,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MaterializationCapability {
    Full,
    Partial,
    Prefetch,
    AsyncFetch,
    FallbackRecompute,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MatchClassification {
    Exact,
    Prefix,
    Partial,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PlanDisposition {
    FullReuse,
    PartialReuse,
    Recompute,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct TensorSpec {
    pub name: String,
    pub role: TensorRole,
    pub dtype: String,
    pub shape: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct QuantizationMetadata {
    pub scheme: String,
    pub bits: u32,
    pub group_size: u32,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct LayoutMetadata {
    pub layout: String,
    pub page_tokens: u32,
    pub block_tokens: u32,
    pub packed: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct TransferPath {
    pub backend: TransferBackend,
    pub capabilities: Vec<TransferCapability>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct MaterializationProfile {
    pub capabilities: Vec<MaterializationCapability>,
    pub tier_kinds: Vec<TierKind>,
    pub device_classes: Vec<DeviceClass>,
    pub buffer_kinds: Vec<BufferKind>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct AttentionStateDescriptor {
    pub schema_version: String,
    pub descriptor_id: String,
    pub engine_family: EngineFamily,
    pub semantic_type: StateSemanticType,
    pub granularity: Granularity,
    pub tensor_specs: Vec<TensorSpec>,
    pub quantization: QuantizationMetadata,
    pub layout: LayoutMetadata,
    pub compatibility_flags: Vec<CompatibilityFlag>,
    pub transfer_paths: Vec<TransferPath>,
    pub materialization: MaterializationProfile,
    pub layout_metadata: BTreeMap<String, String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct KeyIdentity {
    pub tenant: String,
    pub namespace: String,
    pub model: String,
    pub engine_family: EngineFamily,
    pub semantic_type: StateSemanticType,
    pub tokens: Vec<u32>,
    pub block_id: Option<u32>,
    pub page_id: Option<u32>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReuseKey {
    pub identity: KeyIdentity,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct QueryKey {
    pub identity: KeyIdentity,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct EntryVersion {
    pub generation: u32,
    pub lineage: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct EntryLocation {
    pub tier: TierKind,
    pub locator: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct PolicyHint {
    pub reusable: bool,
    pub admission_hint: String,
    pub eviction_hint: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct EntryIdentity {
    pub key: KeyIdentity,
    pub entry_id: String,
    pub version: EntryVersion,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct CacheEntry {
    pub identity: EntryIdentity,
    pub descriptor: AttentionStateDescriptor,
    pub location: EntryLocation,
    pub policy_hint: PolicyHint,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct MatchExtent {
    pub units: u32,
    pub granularity: Granularity,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct RemainingWork {
    pub tokens: Vec<u32>,
    pub fetch_required: bool,
    pub recompute_required: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct CompatibilitySignal {
    pub reusable: bool,
    pub fallback_to_recompute: bool,
    pub reason: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct MatchResult {
    pub classification: MatchClassification,
    pub matched_key: ReuseKey,
    pub requested_key: QueryKey,
    pub matched_extent: MatchExtent,
    pub entry: CacheEntry,
    pub remaining: RemainingWork,
    pub compatibility: CompatibilitySignal,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReusableSlice {
    pub tokens: Vec<u32>,
    pub source_tier: TierKind,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct PartialHitPlan {
    pub disposition: PlanDisposition,
    pub reusable: ReusableSlice,
    pub remaining: RemainingWork,
    pub entry: CacheEntry,
}
